contracts/utils/MultiRateLimited.sol

```diff
@@ -0,0 +1,272 @@
+/// and then there is a global buffer which is the buffer that each individual address must respect as well
+/// @author Fei Protocol
+contract MultiRateLimited is RateLimited, IMultiRateLimited {
+    using SafeCast for *;
+
+    /// @notice the struct containing all information per rate limited address
+    struct RateLimitData {
+        uint32 lastBufferUsedTime;
+        uint144 bufferCap;
+        uint112 rateLimitPerSecond;
+        uint144 bufferStored;
+    }
+
+    /// @notice rate limited address information
+    mapping (address => RateLimitData) public rateLimitPerAddress;
+
+    /// @notice max rate limit per second allowable by non governor
+    uint256 public maxRateLimitPerSecond;
+
+    /// @notice max buffer cap allowable by non governor
+    uint256 public maxBufferCap;
+
+    constructor(
+        address coreAddress,
+        uint256 _maxRateLimitPerSecond,
+        uint256 _rateLimitPerSecond,
+        uint256 _maxRateLimitPerSecondMRL,
+        uint256 _maxBufferCap,
+        uint256 _bufferCap,
+        bool _doPartialAction
+    )
+        CoreRef(coreAddress)
+        RateLimited(_maxRateLimitPerSecond, _rateLimitPerSecond, _bufferCap, _doPartialAction)
+    {
+        require(_maxBufferCap < _bufferCap, "MultiRateLimited: max buffer cap invalid");
+
+        maxRateLimitPerSecond = _maxRateLimitPerSecondMRL;
+        maxBufferCap = _maxBufferCap;
+    }
+
+    // ----------- Governor and Admin only state changing api -----------
+
+    /// @notice update the sub gov rate limit per second
+    /// @param newRateLimitPerSecond new maximum rate limit per second for sub governors
+    /// TODO determine the proper modifier for this function
+    function updateMaxRateLimitPerSecond(uint256 newRateLimitPerSecond) external virtual override onlyGovernor {
+        uint256 oldMaxRateLimitPerSecond = maxRateLimitPerSecond;
+        maxRateLimitPerSecond = newRateLimitPerSecond;
+
+        emit MultiMaxRateLimitPerSecondUpdate(oldMaxRateLimitPerSecond, newRateLimitPerSecond);
+    }
+
+    /// @notice update the sub gov max buffer cap
+    /// @param newSubGovBufferCap new buffer cap for sub governor added addresses
+    /// TODO determine the proper modifier for this function
+    function updateMaxBufferCap(uint256 newSubGovBufferCap) external virtual override onlyGovernor {
+        uint256 oldBufferCap = maxBufferCap;
+        maxBufferCap = newSubGovBufferCap;
+
+        emit MultiBufferCapUpdate(oldBufferCap, newSubGovBufferCap);
+    }
+
+    /// @notice add an authorized rateLimitedAddress contract
+    /// @param rateLimitedAddress the new address to add as a rateLimitedAddress
+    /// @param _rateLimitPerSecond the rate limit per second for this rateLimitedAddress
+    /// @param _bufferCap  the buffer cap for this rateLimitedAddress
+    function addAddress(
+        address rateLimitedAddress,
+        uint112 _rateLimitPerSecond,
+        uint144 _bufferCap
+    ) external virtual override onlyGovernor {
```

Comment:
It's very clean to tuck all of this logic into a helper function. Because the `_addAddress` function is used in multiple places I think this is quite handy and reduces code size and duplication.